App/model links (add only)

nishadsingh1 opened the pull request:
This PR addresses the third checkbox in #218. It only allows for adding links -- I'll work on a PR that allows for removing links after this gets edited and merged.

Given that users cannot yet remove links and that we limit apps to having at most 1 linked model, the only real difference this PR introduces to Clipper users is that app's candidate models (we now call them linked models) can be specified after app registration.

Commit:
First steps

Commit:
Management tests passing

Commit:
Formatted code + fixed breaking tests in non-management suites

Commit:
Added redis and query frontend tests

Commit:
Reformatted

Commit:
Clipper manager now updated

Commit:
Updated python integration tests

Commit:
Renaming, formatting, minor fixes

Commit:
Updated tutorial

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/463/
Test FAILed.

nishadsingh1 commented:
Looks like it's a formatting error: https://amplab.cs.berkeley.edu/jenkins/job/Clipper-PRB/463/consoleFull
"Found 1 C++ style violations.
Run format_code.sh to fix."

My code formatting script doesn't recommend any changes. Would one of you mind pushing up a formatting fix?

dcrankshaw committed:
format code

nishadsingh1 commented:
Thanks @dcrankshaw

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/465/
Test FAILed.

nishadsingh1 commented:
"Build timed out (after 120 minutes). Marking the build as aborted.
Build was aborted" -- is this normal / should we just restart the build?

nishadsingh1 commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/474/
Test FAILed.

Commit:
Changed a few calls to register_application to fit the new signature

Commit:
Updated for easy app registration+model deployment

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/485/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/486/
Test FAILed.

nishadsingh1 commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/487/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/489/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/490/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/491/
Test FAILed.

nishadsingh1 commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/493/
Test FAILed.

nishadsingh1 commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/494/
Test FAILed.

shaneknapp commented:
test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/495/
Test FAILed.

shaneknapp commented:
test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/496/
Test FAILed.

Commit:
Reordered tests + added logging

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/500/
Test FAILed.

Commit:
Reformatted

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/502/
Test PASSed.

nishadsingh1 commented:
Somehow, tests passed this time without any functional changes (just added print statements):
```
[ RUN      ] RedisTest.SubscriptionDetectModelLinksAdd
[12:08:01.529][info]       [REDIS] Successfully issued command "FLUSHALL"
[12:08:01.529][info]       [REDIS] Successfully issued command "CONFIG SET notify-keyspace-events AKE"
[12:08:01.529][info]       [REDIS] Successfully issued command "SELECT 5"
[12:08:01.529][info]       [REDIS] Successfully issued command "HMSET my_app_name input_type doubles policy exp3_policy default_output 1.0 latency_slo_micros 10000"
[12:08:01.529][info]       [REDIS] SUBSCRIPTION STRING: __keyspace@7__:*
[12:08:02.030][info]   [REDISTEST] Before add_mode_links
[12:08:02.030][info]       [REDIS] Successfully issued command "SELECT 7"
[12:08:02.031][info]       [REDIS] Successfully issued command "SADD my_app_name model1"
[12:08:02.031][info]       [REDIS] MESSAGE: sadd
[12:08:02.031][info]   [REDISTEST] Received event
[12:08:02.031][info]   [REDISTEST] Confirmed event is SADD
[12:08:02.031][info]   [REDISTEST] Got lock in subscription callback
[12:08:02.031][info]   [REDISTEST] Before cv notify
[12:08:02.031][info]   [REDISTEST] After cv notify
[12:08:02.031][info]       [REDIS] MESSAGE: sadd
[12:08:02.031][info]   [REDISTEST] Received event
[12:08:02.031][info]   [REDISTEST] Confirmed event is SADD
[12:08:02.031][info]   [REDISTEST] Got lock in subscription callback
[12:08:02.031][info]   [REDISTEST] Before cv notify
[12:08:02.031][info]   [REDISTEST] After cv notify
[12:08:02.031][info]       [REDIS] Successfully issued command "SADD my_app_name model2"
[12:08:02.031][info]   [REDISTEST] After add_mode_links
[12:08:02.031][info]   [REDISTEST] Got lock outside of subscription callback
[12:08:02.031][info]   [REDISTEST] After waiting on cv
[       OK ] RedisTest.SubscriptionDetectModelLinksAdd (1527 ms)
```

Maybe I'll wait on getting this PR into its final format before worrying about why Jenkins tests are failing/passing on this code (since these occasional failures seem to be specific to this PR)

Commit:
Revert of test reordering and logging

nishadsingh1 commented on src/libclipper/src/redis.cpp:
```diff
@@ -438,6 +449,22 @@ bool add_application(redox::Redox& redis, const std::string& appname,
+                     const std::vector<std::string>& model_names) {
+  if (send_cmd_no_reply<string>(
+          redis, {"SELECT", std::to_string(REDIS_APP_MODEL_LINKS_DB_NUM)})) {
+    for (auto model_name : model_names) {
```
We're making multiple requests to Redis here instead of sending the whole set of `model_names` at once. This is because I couldn't find a way to add multiple elements to a set using the available `send_cmd_no_reply` infrastructure in one call. Let me know if there's a way to do this.

It's possible that no such way exists and sending a request for each new linked model is a bad practice. In this case, we could just have this REDIS_APP_MODEL_LINKS table store a mapping between app names and a string. The string would be a serialized list of model names (maybe with one of our predefined list delimiters). Then, whenever we want to fetch model links, we'd just deserialize the list. When adding new ones, we'd fetch the list, deserialize it, add to it, and re-serialize, and send it to Redis.

For now it's possible this isn't a big deal because we enforce that `model_names` has no more than 1 entry.

Corey-Zumar replied:
@dcrankshaw Any advice here?

It's true that this is fine for now because of the restriction that an application may only have one backing model. However, it would be nice to write this efficiently for future use cases when the restriction is loosened.

Corey-Zumar replied:
Resolved. This function is not on any critical paths, so making a few extra database queries will not have any significant impact on performance.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/507/
Test FAILed.

Corey-Zumar commented on README.md:
```diff
@@ -46,16 +46,20 @@ Checking if Docker is running...
-# http://localhost:1337/hello_world/predict
+# Register an application called "hello_world". This will create
+# a prediction REST endpoint at http://localhost:1337/hello_world/predict
 >>> clipper.register_application("hello_world", "feature_sum_model", "doubles", "-1.0", 100000)
```
This no longer conforms to the method prototype of `register_application`. We should remove `"feature_sum_model"`.

Corey-Zumar commented on clipper_admin/clipper_manager.py:
```diff
@@ -371,6 +368,52 @@ def register_application(self, name, model, input_type, default_output,
         return r.status_code == requests.codes.ok
 
+    def link_model_to_app(self, app_name, model_name):
+        """Allows the model to be used by app
```
How does "Allows the model with `model_name` to be used by the app with `app_name`" sound?

We should also note that `app_name` should be a preexisting application that has already been registered with clipper, while the model does not have to be registered.

Corey-Zumar commented on clipper_admin/clipper_manager.py:
```diff
@@ -371,6 +368,52 @@ def register_application(self, name, model, input_type, default_output,
+        })
+        headers = {'Content-type': 'application/json'}
+        r = requests.post(url, headers=headers, data=req_json)
+        print(r.text)
```
Remove extraneous print statement

Corey-Zumar commented on src/management/src/management_frontend.hpp:
```diff
@@ -414,13 +472,89 @@ class RequestHandler {
+        return "Success!";
+      } else {
+        std::stringstream ss;
+        ss << "There already exist links to different models for app "
```
Nit: `A model with name <model_name> is already linked to <app_name>` is cleaner.

Corey-Zumar commented on clipper_admin/clipper_manager.py:
```diff
@@ -371,6 +368,52 @@ def register_application(self, name, model, input_type, default_output,
+        return r.status_code == requests.codes.ok
+
+    def get_linked_models(self, app_name):
+        """Retrieves the models linked to the app
```
Replace "app" with "specified application".

Corey-Zumar commented on examples/tutorial/tutorial_part_one.ipynb:
```diff
@@ -220,14 +220,12 @@
-    "model_name = \"birds_vs_planes_classifier\"\n",
-    "# If the model doesn't return a prediction in time, predict\n",
-    "# label 0 (bird) by default\n",
+    "# If the model (which we will later link to our apo) doesn't\n",
```
Nit: typo. `apo` >> `application`

Corey-Zumar commented on examples/tutorial/tutorial_part_one.ipynb:
```diff
@@ -253,6 +251,52 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Clipper applications don't generate predictions themselves; models do. Let Clipper know that your `\"cifar_demo\"` app should use the `\"birds_vs_planes_classifier\"` model to serve predictions once it is deployed.\n",
```
"In order to generate predictions, a Clipper application needs to be linked to a model. Let Clipper know that your..." is clearer.

Corey-Zumar commented on examples/tutorial/tutorial_part_one.ipynb:
```diff
@@ -197,11 +197,11 @@
     "# Create an application\n",
     "\n",
-    "In order to query Clipper for predictions, you need to create an application. Each application specifies a name, a set of models it can query, the query input datatype, the selection policy, and a latency service level objective. Once you register an application with Clipper, the system will create two REST endpoints: one for requesting predictions and for providing feedback.\n",
+    "In order to query Clipper for predictions, you need to create an application. Each application specifies a name, the query input datatype, the selection policy, and a latency service level objective. Once you register an application with Clipper, the system will create two REST endpoints: one for requesting predictions and for providing feedback.\n",
```
While we're here, let's remove the reference to the "feedback" endpoint, as we don't currently support it or incorporate it as part of the tutorial.

Corey-Zumar commented on examples/tutorial/tutorial_part_one.ipynb:
```diff
@@ -253,6 +251,52 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "You can view which models your app is linked to by running the below code."
```
Nit: `below code` >> `code below`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/516/
Test FAILed.

Corey-Zumar commented on src/frontends/src/query_frontend.hpp:
```diff
@@ -156,16 +156,34 @@ class RequestHandler {
+                                        "New model link detected for app: {}",
+                                        app_name);
+
+            boost::optional<std::vector<std::string>> linked_model_names =
```
Why do we reference `linked_model_names` as an`optional` here when `clipper::redis::get_linked_models` returns a vector?

Corey-Zumar commented on src/frontends/src/query_frontend.hpp:
```diff
@@ -156,16 +156,34 @@ class RequestHandler {
+
+            boost::optional<std::vector<std::string>> linked_model_names =
+                clipper::redis::get_linked_models(redis_connection_, app_name);
+            if (linked_model_names) {
```
If `linked_model_names` does not need to be an optional, we can instead check whether or not it's empty.

Corey-Zumar replied:
This method may not be necessary (see my comment below about whether or not it's a good idea to store link information locally)

Corey-Zumar commented on src/libclipper/include/clipper/redis.hpp:
```diff
@@ -182,6 +180,14 @@ std::vector<std::string> get_all_model_names(redox::Redox& redis);
 /**
+ * Looks up which models are linked to app with name `app_name`
+ * \return Returns a vector of model names. If no models are linked
+ * to the specified app, then am empty vector will be returned.
```
Nit: Typo `am` >> `an`

Corey-Zumar commented on clipper_admin/clipper_manager.py:
```diff
@@ -371,6 +368,52 @@ def register_application(self, name, model, input_type, default_output,
+        print(r.text)
+        return r.status_code == requests.codes.ok
+
+    def get_linked_models(self, app_name):
```
This results in `ValueError: No JSON object could be decoded` in the case that `app_name` is not a registered application.

Corey-Zumar commented on src/frontends/src/query_frontend.hpp:
```diff
@@ -248,9 +268,20 @@ class RequestHandler {
+    linked_models_for_apps_[name] = models;
+  }
+
+  std::vector<std::string> get_linked_models_for_app(std::string name) {
```
What's the rationale behind storing a copy of `linked_models_for_apps_` rather than querying Redis when a request is made for the data? These kinds of queries are probably infrequent and won't present an excessive load on the database. Unless there's a good reason, we should rely on Redis as a source of truth.

Corey-Zumar commented on src/libclipper/src/json_util.cpp:
```diff
@@ -384,6 +384,33 @@ std::string to_json_string(rapidjson::Document& d) {
+void set_string_array(rapidjson::Document& d,
+                      const std::vector<std::string>& str_array) {
+  d.SetArray();
+  for (auto str : str_array) {
```
This should be `for (auto const &str: str_array) {`. The reason being that using `auto` alone iterates by value, which I'm pretty sure incurs a copy.

Corey-Zumar commented on src/libclipper/include/clipper/redis.hpp:
```diff
@@ -251,12 +257,20 @@ std::vector<std::pair<VersionedModelId, int>> get_all_containers(
+ *
+ * \return Returns true if the add was successful.
+ */
+bool add_model_links(redox::Redox& redis, const std::string& appname,
```
Nit: `appname` >> `app_name`

Corey-Zumar commented on src/libclipper/src/redis.cpp:
```diff
@@ -209,6 +205,24 @@ boost::optional<std::string> get_current_model_version(
+          redis, {"SELECT", std::to_string(REDIS_APP_MODEL_LINKS_DB_NUM)})) {
+    auto result =
+        send_cmd_with_reply<std::vector<string>>(redis, {"SMEMBERS", app_name});
+    linked_models = *result;
```
We should check that `result` is present before attempting to dereference it from the `optional` container.

Corey-Zumar commented on src/management/src/management_frontend.hpp:
```diff
@@ -64,13 +68,25 @@ const std::string GET_CONTAINER = ADMIN_PATH + "/get_container$";
+const std::string ADD_MODEL_LINKS_JSON_SCHEMA = R"(
+  {
+    "app_name" := string,
+    "model_names" := := [string]
```
Nit: redundant `:=`

Corey-Zumar reviewed (changes requested):
Functionality is almost complete! I really like the new link interface. Most of my comments are cosmetic.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/519/
Test FAILed.

nishadsingh1 commented:
- [x] Make sure models exist before linking them

Commit:
Addressed changes Corey requested, made minor changes in clipper manager's documentation, models now need to be registered before getting linked

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/523/
Test FAILed.

nishadsingh1 commented:
Thanks @Corey-Zumar for the review -- the latest commit should address your comments and enforce that models are registered with Clipper before linking them to any apps.

Tests are passing locally for me; they're failing on Jeknins because the spark scala container isn't building correctly (fix attempt at #248), though of course there may be legitimate failures too that will show up later.

Even without the passing Jenkins tests, I think this is still ready for another pass-over.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/529/
Test FAILed.

Commit:
Added print statements to hanging RedisTest.SubscriptionDetectModelLinksAdd

nishadsingh1 commented:
Looks like it's hanging on `RedisTest. SubscriptionDetectModelLinksAdd`, even when print statements are added back in.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/530/
Test FAILed.

shaneknapp commented:
i just checked the worker that this test ran on and it's got a TON zombie
processes.  this could be causing issues.

i'll need to reboot this machine to get rid of them.

On Fri, Jul 7, 2017 at 10:30 AM, nishadsingh1 <notifications@github.com>
wrote:

> Looks like it's hanging on RedisTest. SubscriptionDetectModelLinksAdd,
> even when print statements are added back in.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ucbrise/clipper/pull/226#issuecomment-313744735>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABiDrK4UMd9aFpIfCE4H0ENh0h9Z-Q7tks5sLmswgaJpZM4OCFq7>
> .
>

Commit:
Updated tests and documentation to reflect that models need to be registered before linking

Commit:
Added some tests + got them passing

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/533/
Test FAILed.

Corey-Zumar commented on clipper_admin/clipper_manager.py:
```diff
@@ -371,6 +368,58 @@ def register_application(self, name, model, input_type, default_output,
+            "model_names": [model_name]
+        })
+        headers = {'Content-type': 'application/json'}
+        r = requests.post(url, headers=headers, data=req_json)
```
On second thought, let's re-add the call to `print(r.text)`. It's very helpful for user understanding - sorry!

nishadsingh1 replied:
Np at all

Corey-Zumar commented on src/libclipper/src/redis.cpp:
```diff
@@ -438,6 +454,22 @@ bool add_application(redox::Redox& redis, const std::string& appname,
   }
 }
 
+bool add_model_links(redox::Redox& redis, const std::string& appname,
```
Currently, this method allows users to link an application of input type A to a model of input type B. We should verify that the input types are the same and fail if they aren't. Let's add a test for this as well.

nishadsingh1 replied:
Great catch :) I think it's correct to do this in `redis.cpp`, but we could also do it in the management front end. Just want to confirm that `redis.cpp` is the right place for it - thoughts?

nishadsingh1 replied:
Actually, I'm going to redistribute the validation responsibilities.
Management frontend:
- Confirm that the user supplied only one model_name
- Make sure that there will only be one link

Redis:
- Confirm that an app with specified app name exists
- Confirm that a model exists for each specified model name
- Confirm input types match

It seems like we'd want the checks that guarantee important logic to exist at the single entry point to the database (redis.cpp). Let me know if that sounds reasonable.

--------------------
Edit:  It looks like we aren't propagating errors up from any of the functions in `redis.cpp`, so there isn't a standard way to handle them (and access their specific error messages) from the management frontend function.

We should definitely still let the user know when their specified app/model doesn't exist or if the input types don't match, instead of returning some nondescriptive error message in all cases. In order to give the appropriate error message, we need the management frontend function to make the checks that are under the 'Redis' section above.

Should we just duplicate the checks like so?:

Management frontend:
- Confirm that the user supplied only one model_name
- Make sure that there will only be one link
- Confirm that an app with specified app name exists
- Confirm that a model exists for each specified model name
- Confirm input types match

Redis:
- Confirm that an app with specified app name exists
- Confirm that a model exists for each specified model name
- Confirm input types match

I think this is probably okay since adding model links isn't on a critical path.

--------------------
Edit 2: We don't seem to do much validation at all in `redis.cpp`. Is this intentional / does this mean that we should only do this validation in the management frontend?

nishadsingh1 replied:
We resolved to just have all the checks be in the Management Frontend and to move them to Redis in another PR

Corey-Zumar reviewed (changes requested):
Almost there - just missing some input type verification during the linking process

nishadsingh1 commented:
**Problematic scenario 1:**
- Register a model with name 'M', version '1', and input_type 'ints'.
- Register an app with name 'A' and input_type 'ints'
- Add a model link between 'M' and 'A'. Input types match, both apps exist, only one model name was specified and the app had no previous links: this request is successful.
- Register a model with name 'M', version '2', and input_type 'doubles'.

Queries to 'A' route to the latest version of 'M'. In this case, that would send requests to 'M:2', which has an incompatible input_type. Under the hood, registering 'M:2' set the current version of 'M' to '2'. This was the source of the problem.

**Problematic scenario 2**
- Register a model with name 'M', version '1', and input_type 'ints'.
- Register a model with name 'M', version '2', and input_type 'doubles'.
- Register an app with name 'A' and input_type 'doubles'
- Add a model link between 'M' and 'A'. Input types match, both apps exist, only one model name was specified and the app had no previous links: this request is successful.
- Set the current model version to '1'. 

Queries to 'A' route to 'M:1' which has an incompatible input_type.

So, it seems like we need to make sure that every time we set the current model version of a model, we're not creating some conflict with an existing linked application with respect to input type.

**Where to do this validation?**
If we put it in management frontend, we'd need to do the check in both `add_model` and `set_model_version`.

If we put it in redis.cpp's, we'd only need to do the check in `set_current_model_version`.

I like the latter option. If every time we change a model's model version, shouldn't we just have the check be in the function (or on a guaranteed path to the function ) that makes the change? Otherwise we run the risk of forgetting to run these checks with each usage of redis' `set_current_model_version` and creating bugs.

------
Update: we'll have all checks be in Management Frontend for now and move them to Redis in another PR

Commit:
Now checks for input type compatibility

Commit:
Updated frontend set model version code (no logical changes)

nishadsingh1 commented:
Updates on the latest change:
- Addressed Corey's comments
- Made sure that input types matched for new links
- Made sure that when adding a new model, there exist no previous versions were linked to apps under different input types
- Made sure that when setting a new model version, there exist no links to applications under a the current model's input type that would be incompatible under the version change
- In the process of the latter two points, rewrote the frontend set_model_version code to match the form of our other frontend endpoints (plus, the change was necessary to avoid some uglier code changes to accommodate for the above checks)

Commit:
Reverting strange tutorial cell status changes

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/540/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/542/
Test FAILed.

Corey-Zumar commented on src/management/src/management_frontend.hpp:
```diff
@@ -511,23 +643,52 @@ class RequestHandler {
       throw std::invalid_argument(ss.str());
     }
+
+    // Need to fux with this
```
Remove comment. Are there outstanding changes that need to be made?

Corey-Zumar commented on src/management/src/management_frontend.hpp:
```diff
@@ -836,8 +1042,35 @@ class RequestHandler {
+    std::string model_name = get_string(d, "model_name");
+    std::string new_model_version = get_string(d, "model_version");
+
+    check_updated_model_consistent_with_app_links(model_name,
```
The `check_updated_model_consistent_with_app_links` method is being invoked with `new_model_version` as the second parameter, as opposed to an `input_type` value.

Also, `attempt_model_version_update` calls the same method again. Does it need to be called in both places?

Corey-Zumar reviewed (changes requested):
one more issue & doc cleanup. ready to merge when these changes are made and tests pass.

Commit:
Addressed comments

Commit:
Set model version throws error if model doesn't exist, added test. More print statements in hanging test.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/543/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/544/
Test FAILed.

Commit:
Fixed hanging redis add model links test

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/545/
Test FAILed.

Commit:
Models should now link correctly in pyspark and R integration tests

Corey-Zumar reviewed (approved):
Looks good. Will merge when tests pass. Looks like there's an issue with pyspark model deployment.

dcrankshaw commented:
Actually hold off on merging this. This is not backwards compatible with the 0.1 release and will cause problems due to #244

dcrankshaw commented:
Actually you can go ahead and merge it. I turned off the automated builds on docker hub until we can address #244. I'll prioritize that and push out a new 0.1.x release that manages docker versions properly.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/546/
Test FAILed.

Commit:
Updated many apps many models test to link correctly

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/549/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/550/
Test PASSed.
